src/snippet/mod.rs

```diff
@@ -0,0 +1,207 @@
+                          stop_offset: end_offset,
+                          num_chars: 0,
+                          highlighted: vec![]}
+    }
+
+    /// Updates `score` and `highlighted` fields of the objects.
+    ///
+    ///
+    fn calculate_score(&mut self, token: &Token, terms: &BTreeMap<String, f32>) {
+        if let Some(score) = terms.get(&token.text.to_lowercase()) {
+            self.score += score;
+            self.highlighted.push(HighlightSection{start: token.offset_from,
+                                                   stop: token.offset_to});
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Snippet {
+    fragments: String,
+    highlighted: Vec<HighlightSection>,
+}
+
+const HIGHLIGHTEN_PREFIX:&str = "<b>";
+const HIGHLIGHTEN_POSTFIX:&str = "</b>";
+
+impl Snippet {
+
+    /// Returns a hignlightned html from the `Snippet`.
+    pub fn to_html(&self) -> String {
+        let mut html = String::new();
+        let mut start_from: usize = 0;
+
+        for item in self.highlighted.iter() {
+            html.push_str(&self.fragments[start_from..item.start]);
+            html.push_str(HIGHLIGHTEN_PREFIX);
+            html.push_str(&self.fragments[item.start..item.stop]);
+            html.push_str(HIGHLIGHTEN_POSTFIX);
+            start_from = item.stop;
+        }
+        html.push_str(&self.fragments[start_from..self.fragments.len()]);
+        html
+    }
+}
+
+/// Returns a non-empty list of "good" fragments.
+///
+/// If no target term is within the text, then the function
+/// should return an empty Vec.
+///
+/// If a target term is within the text, then the returned
+/// list is required to be non-empty.
+///
+/// The returned list is non-empty and contain less
+/// than 12 possibly overlapping fragments.
+///
+/// All fragments should contain at least one target term
+/// and have at most `max_num_chars` characters (not bytes).
+///
+/// It is ok to emit non-overlapping fragments, for instance,
+/// one short and one long containing the same keyword, in order
+/// to leave optimization opportunity to the fragment selector
+/// upstream.
+///
+/// Fragments must be valid in the sense that `&text[fragment.start..fragment.stop]`\
+/// has to be a valid string.
+fn search_fragments<'a>(
+    tokenizer: Box<BoxedTokenizer>,
+    text: &'a str,
+    terms: BTreeMap<String, f32>,
+    max_num_chars: usize) -> Vec<FragmentCandidate> {
+    let mut token_stream = tokenizer.token_stream(text);
+    let mut fragment = FragmentCandidate::new(0, 0);
+    let mut fragments:Vec<FragmentCandidate> = vec![];
+
+    loop {
+        if let Some(next) = token_stream.next() {
+            if (next.offset_to - fragment.start_offset) > max_num_chars {
+                let txt = &text[fragment.start_offset..fragment.stop_offset];
+                if fragment.score > 0.0 {
+                    fragments.push(fragment)
+                };
+                fragment = FragmentCandidate::new(next.offset_from, next.offset_to);
```

Comment:
I think there is a bug if `next` is a keyword. Then it never contributes to the score does it.